src/parsing.rs

```diff
@@ -106,14 +111,31 @@ fn read_n_bytes<R: Read>(file: &mut R, length: usize) -> io::Result<Vec<u8>> {
 
 #[allow(clippy::vec_box)]
 fn read_constant_pool<R: Read>(file: &mut R) -> io::Result<Vec<Box<ConstantPoolEntry>>> {
-    let constant_pool_count = i32::from(read_u16(file)?);
+    let mut constant_pool_count = read_u16(file)? - 1;
 
-    let mut constant_pool = Vec::<Box<ConstantPoolEntry>>::new();
+    let mut constant_pool = Vec::<Box<ConstantPoolEntry>>::with_capacity(constant_pool_count as usize);
```

Comment:
Is it guaranteed that the constant pool will need this many elements? Will longs and doubles only require 1 space in the constant pool, but account for 2 spaces in the `constant_pool_count` at this point?

Reply:
here is the relevant snippet from jvm spec:

> The value of the constant_pool_count item is equal to the number of entries in the constant_pool table plus one. A constant_pool index is considered valid if it is greater than zero and less than constant_pool_count, with the exception for constants of type long and double noted in §4.4.5.

so the constant_pool_count is the size of the table, which, with a dense structure like Vec, must be large enough to contain every addressable index in the table. other option is to use a mapping table or a HashMap but i think simplicity of Vec wins here

Reply:
Ah yep, you're right.

I found this section of the specification convincing on this point:

> All 8-byte constants take up two entries in the constant_pool table of the class file. If a CONSTANT_Long_info or CONSTANT_Double_info structure is the entry at index n in the constant_pool table, then the next usable entry in the table is located at index n+2. The constant_pool index n+1 must be valid but is considered unusable.

From: https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.4.5

In particular:

> The constant_pool index n+1 must be valid but is considered unusable.

Reply:
yep, i dropped exactly that snipet into a code comment